tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,284 @@
+    model = tf.keras.Sequential([
+        # The input is the data-circuit, encoded as a tf.string
+        tf.keras.layers.Input(shape=(), dtype=tf.string),
+        # The PQC layer returns the expected value of the
+        # readout gate, range [-1,1].
+        tfq.layers.PQC(circuit, cirq.Z(q1)),
+    ])
+
+    def hinge_loss(y_true, y_pred):
+        # Here we use hinge loss as the cost function
+        tf.reduce_mean(tf.cast(1 - y_true * y_pred, tf.float32))
+
+    result = rotosolve_minimizer.minimize(
+        rotosolve_minimizer.function_factory(
+            model, hinge_loss, x_circ, Y), np.random.rand([2])
+            ) # Initial guess of the parameter
+
+    print(result)
+    ```
+
+    Args:
+      expectation_value_function:  A Python callable that accepts
+        a point as a real `Tensor` and returns a tuple of `Tensor`s
+        of real dtype containing the value of the function and its
+        gradient at that point. The function to be minimized. The
+        input is of shape `[..., n]`, where `n` is the size of the
+        domain of input points, and all others are batching dimensions.
+        The first component of the return value is a real `Tensor` of
+        matching shape `[...]`. The second component (the gradient)
+        is also of shape `[..., n]` like the input value to the function.
+        This must be a linear combination of quantum measurement
+        expectation value, otherwise this algorithm cannot work.
+      initial_position: Real `Tensor` of shape `[..., n]`. The starting
+        point, or points when using batching dimensions, of the search
+        procedure. At these points the function value and the gradient
+         norm should be finite.
+      tolerance: Scalar `Tensor` of real dtype. Specifies the gradient
+        tolerance for the procedure. If the supremum norm of the gradient
+        vector is below this number, the algorithm is stopped.
+      name: (Optional) Python str. The name prefixed to the ops created
+        by this function. If not supplied, the default name 'minimize'
+        is used.
+
+    Returns:
+      optimizer_results: A RotosolveOptimizerResults object contains the
+       result of the optimization process.
+    """
+
+    with tf.name_scope(name or 'minimize'):
+        initial_position = tf.convert_to_tensor(initial_position,
+                                                name='initial_position',
+                                                dtype='float32')
+        dtype = initial_position.dtype.base_dtype
+        tolerance = tf.convert_to_tensor(tolerance,
+                                         dtype=dtype,
+                                         name='grad_tolerance')
+        max_iterations = tf.convert_to_tensor(max_iterations,
+                                              name='max_iterations')
+
+        def _rotosolve_one_parameter_once(state):
+            """
+            Rotosolve a single parameter.
+            """

```

Comment:
1) This internal function is hard to test. could you bring it out and write a test code?
2) one line description + arguments description please.

Reply:
Well, this function is written here purely to help people read it, and it's a bit difficult to extract it out since too many variables are used in this closure. Also, the rest of the code is nothing but iterate over parameters and calling this function. Do you think it is okay to leave it as an internal function?

Reply:
Docstring is fixed in 64f76c59d428d1501648637c7f1cc3af399cc7ed